qualtran/surface_code/rotation_cost_model.py

```diff
@@ -0,0 +1,109 @@
+
+    @abc.abstractmethod
+    def prepartion_overhead(self, error_budget) -> AlgorithmSummary:
+        """Cost of preparation circuit."""
+
+
+@frozen
+class LogarithmicModel(RotationCostModel):
+    r"""A linear model in the log of the error budget with no preparation cost.
+
+    $$
+    \#T = -\textit{slope} \log_2{\textit{budget}} + \textit{overhead}
+    $$
+
+    Attributes:
+        slope: The coefficient of $log_2{budget}$.
+        overhead: The overhead.
+        gateset: A human-readable description of the gate set (e.g. 'Clifford+T').
+        approximation_protocol: A description or reference to the approximation protocol (e.g. `Diagonal:https://arxiv.org/abs/1403.2975`)
+        reference: A human-readable description of the source of the model (e.g. 'https://arxiv.org/abs/1404.5320')
+    """
+    slope: float
+    overhead: float
+    gateset: str | None = None
+    approximation_protocol: str | None = None
+    reference: str | None = None
+
+    def rotation_cost(self, error_budget: float) -> AlgorithmSummary:
+        return AlgorithmSummary(
+            t_gates=math.ceil(-self.slope * math.log2(error_budget) + self.overhead)
+        )
+
+    def prepartion_overhead(self, error_budget) -> AlgorithmSummary:
+        return AlgorithmSummary()
+
+
+@frozen
+class ConstantCostWithPreparation(RotationCostModel):
+    r"""A rotation cost of bitsize-2 toffoli per rotation independent of the error budget.
+
+    This model assumes a state $\ket{\phi}$ has been prepared using a standard technique, then
+    each rotation is applied with bitsize digits of accuracy using bitsize-2 Toffoli gates.
+    $$
+    \ket{\phi} = \frac{1}{\sqrt{2^{b}}} \sum_{k=0}^{2^b-1} e^{-2\pi i k/2^b} \ket{k}
+    $$
+    Where $b$ is the bitsize/number of digits of accuracy.
+
+    reference: https://doi.org/10.1103/PRXQuantum.1.020312
+
+    Attributes:
+        bitsize: Number of digits of accuracy for approximating a rotation.
+        slope: The coefficient of $log_2{budget}$.
+        overhead: The overhead.
+        gateset: A human-readable description of the gate set (e.g. 'Clifford+T').
+        approximation_protocol: A description or reference to the approximation protocol (e.g. `Diagonal:https://arxiv.org/abs/1403.2975`)
+        reference: A human-readable description of the source of the model (e.g. 'https://arxiv.org/abs/1404.5320')
+    """
+
+    bitsize: int
+    slope: float
+    overhead: float
+    gateset: str | None = None
+    approximation_protocol: str | None = None
+    reference: str | None = None
+
+    def rotation_cost(self, error_budget: float) -> AlgorithmSummary:
+        return AlgorithmSummary(toffoli_gates=max(self.bitsize - 2, 0))
+
+    def prepartion_overhead(self, error_budget) -> AlgorithmSummary:
+        return self.bitsize * AlgorithmSummary(
+            t_gates=math.ceil(-self.slope * math.log2(error_budget / self.bitsize) + self.overhead)
+        )
+
+
+BeverlandEtAl = LogarithmicModel(
```

Comment:
Outside of this definition, this variable will appear without the context that it's a rotation cost model

Reply:
updated